osh/hub/scripts/osh-resalloc

```diff
@@ -0,0 +1,159 @@
+    # Copy worker.conf to centos user's home directory
+    # This requires the hub to have ssh access to the new worker
+    logger.debug(f"scp -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o ConnectTimeout=20 -i {settings.RESALLOC_WORKER_SSH_PRIVATE_KEY} {temporary_file.name} centos@{kobo_worker.name}:worker.conf")
+    scp_exit_status = os.system(f"scp -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o ConnectTimeout=20 -i {settings.RESALLOC_WORKER_SSH_PRIVATE_KEY} {temporary_file.name} centos@{kobo_worker.name}:worker.conf")
+    try:
+        if scp_exit_status != 0:
+            logger.error(f"Failed to copy worker.conf to new worker {kobo_worker.name}")
+            return
+    finally:
+        # Delete the worker configs from hub
+        temporary_file.close()
+        # TODO: Shall we delete the worker here?
+
+    logger.debug(f"ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o ConnectTimeout=20 -i {settings.RESALLOC_WORKER_SSH_PRIVATE_KEY} centos@{kobo_worker.name} 'sudo mv worker.conf /etc/osh/worker.conf && sudo chown root:root /etc/osh/worker.conf'")
+    # /etc/osh/hub/id_rsa.worker
+    move_worker_conf_status = os.system(f"ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o ConnectTimeout=20 -i {settings.RESALLOC_WORKER_SSH_PRIVATE_KEY} centos@{kobo_worker.name} 'sudo mv worker.conf /etc/osh/worker.conf && sudo chown root:root /etc/osh/worker.conf'")
+    if move_worker_conf_status != 0:
+        logger.error(f"Failed to move worker.conf to new worker {kobo_worker.name}")
+        # TODO: Shall we delete the worker here?
+        return
+    # Start the worker after this is finished
+    logger.debug(f"ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o ConnectTimeout=20 -i {settings.RESALLOC_WORKER_SSH_PRIVATE_KEY} centos@{kobo_worker.name} 'sudo systemctl restart osh-worker'")
+    start_worker_status = os.system(f"ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o ConnectTimeout=20 -i {settings.RESALLOC_WORKER_SSH_PRIVATE_KEY} centos@{kobo_worker.name} 'sudo systemctl restart osh-worker'")
+    if start_worker_status != 0:
+        logger.error(f"Failed to start worker worker at {kobo_worker.name}")
+        # TODO: Shall we delete the worker here?
+        return
+
+
+def create_worker(ip_address):
+    # Worker name should be set to hostname of the worker
+    worker_name = get_hostname_from_ip(ip_address)
+    kobo_worker = create_worker_in_kobo(worker_name)
+    start_worker(kobo_worker)
+
+
+def delete_worker(ip_address):
+    from kobo.hub import models
+
+    # Worker name should be set to hostname of the worker
+    worker_name = get_hostname_from_ip(ip_address)
+    kobo_worker = models.Worker.objects.get(name=worker_name)
+    kobo_worker.delete()
+
+
+def workers_needed():
+    # This number should not exceed `max` field in the `/etc/resallocserver/pools.yaml`
+    # on the resalloc server.
+    MAX_SINGLE_USE_WORKERS = 8
```

Comment:
This should be moved to the settings file.

Reply:
Resalloc has 3 different values to determine number of virtual machines being spun up in [pools.yaml](https://github.com/praiskup/resalloc/blob/337b16d322d11c524d5e4230ed6c91fd77a1f7f4/config/pools.yaml#L5-L22):

- `max` determines number of maximum virtual machines.
- `max_starting` determines number of maximum virtual machines which are starting.
- `max_prealloc` determines number of maximum pre allocated virtual machines.

Shall we find a way to pick up one of those values to determine number of maximum single use workers? If it's not possible we need to leave a comment that `MAX_SINGLE_USE_WORKERS` should never be above `max` number of nodes from the `pools.yaml`, otherwise the agent spawner daemon would be stuck waiting for tickets that may never be resolved and would not clean up workers that have been already used.

Reply:
> That sounds wrong. Inability to allocate more machines (for whatever reason) should not block releasing of machines that become unused. Otherwise the solution is prone to deadlock.

@praiskup I remember that was the limitation of the first prototype that we did. Does that limitation still exist? The code may have changed over time.

Reply:
I am sure I do not fully understand the question, so just "trying" to answer :-)

- agent spawner just tries to converge to the ideal number of agents, IOW it opens as many resalloc tickets as necessary (plus runs scripts against them, once allocated, if configured to do so)
- the fact that the ticket is open doesn't block the spawner logic, neither resalloc, nor osh
- resalloc's pools.yaml `max` value should be >= the `converge` number, just to be able to converge at some point ... but nothing serious happens even if not (the resalloc tickets stay open..., no more resources are allocated, and agent spawner doesn't receive the agents)

Said that, no blocking (anymore?). All the tasks done by agent spawner are async.

I'm not sure why we need `MAX_SINGLE_USE_WORKERS` here, not sure whether we have to "divide" or categorize OSH workers like that at all. The resalloc workers (agents) may be reused, same as any other workers.  Whether they are/are not is a matter of configuring the agent's hooks appropriately (so ideally, OSH doesn't care at all).

Reply:
> the fact that the ticket is open doesn't block the spawner logic, neither resalloc, nor osh

It was initially blocking in the first prototype we did few months ago. I am happy that it has changed.

> I'm not sure why we need `MAX_SINGLE_USE_WORKERS` here, not sure whether we have to "divide" or categorize OSH workers like that at all. The resalloc workers (agents) may be reused, same as any other workers. Whether they are/are not is a matter of configuring the agent's hooks appropriately (so ideally, OSH doesn't care at all).

It may just be a method to limit number of open tickets in resalloc. For example, what happens if a mass scan is initiated for new version of Fedora. However, it would be good to keep this number in sync with `max` value in `pools.yaml`.